### ISSUE TITLE
tcp-check no longer logs connection failure (step 1)

### ISSUE THREAD

**Issue by lukastribus** [haproxy/haproxy#73]
## Output of `haproxy -vv` and `uname -a`

```
~/haproxy$ ./haproxy -vv
HA-Proxy version 2.0-dev2-9ef2ad-63 2019/04/03 - https://haproxy.org/
Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits
  OPTIONS =

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL -GETADDRINFO -OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 -ZLIB -SLZ +CPU_AFFINITY -TFO -NS +DL +RT -DEVICEATLAS -51DEGREES -SYSTEMD -OBSOLETE_LINKER

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without compression support (neither USE_ZLIB nor USE_SLZ are set).
Compression algorithms supported : identity("identity")
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE
              h2 : mode=HTTP       side=FE
       <default> : mode=HTX        side=FE|BE
       <default> : mode=TCP|HTTP   side=FE|BE

Available services : none

Available filters :
        [SPOE] spoe
        [COMP] compression
        [CACHE] cache
        [TRACE] trace

~/haproxy$ uname -a
Linux dev 4.4.0-109-generic #132-Ubuntu SMP Tue Jan 9 19:52:39 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
~/haproxy$

```

## What's the configuration?

```
global
 daemon
 maxconn 200
defaults
 log global
 mode tcp
 option  tcplog
 timeout connect 120s
 timeout client 120s
 timeout server 120s
frontend ft_redis
  bind :8080
  mode tcp
  default_backend back
backend back
  mode tcp
  option tcp-check
  tcp-check connect port 81 # avoid bug #72
  tcp-check send PING\r\n
  tcp-check expect string +PONG
  server www1 haproxy.org:81 check inter 1s

```

## Steps to reproduce the behavior

1. tcp-check a backend server where the TCP connection is rejected

## Actual behavior

Haproxy logs:
`Connection refused at step 2 of tcp-check (send)`

## Expected behavior

Haproxy should log a connection rejection at step 1 instead:
`Connection refused at step 1 of tcp-check (connect port 81)`

## Analysis

Since @bedis commit 248f1173f ("BUG/MEDIUM: tcp-check: single connect rule can't detect DOWN servers"), we do not wait for the connect() call to return successfully, we immediately pass to step nr 2.

Therefor, when the TCP connection is rejected by the peer, instead of logging:
`Connection refused at step 1 of tcp-check (connect port 81)`

we now see:
`Connection refused at step 2 of tcp-check (send)
`

Stracing after this commit shows that we sendto() the first check immediately after connect() returns EINPROGRESS:

```
22:58:40.879974 socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 7
22:58:40.880087 fcntl(7, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
22:58:40.880176 setsockopt(7, SOL_TCP, TCP_NODELAY, [1], 4) = 0
22:58:40.880285 setsockopt(7, SOL_TCP, TCP_QUICKACK, [0], 4) = 0
22:58:40.880378 connect(7, {sa_family=AF_INET, sin_port=htons(81), sin_addr=inet_addr("51.15.8.218")}, 16) = -1 EINPROGRESS (Operation now in progress)
22:58:40.880492 sendto(7, "PING\r\n", 6, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = -1 EAGAIN (Resource temporarily unavailable)
22:58:40.880605 epoll_ctl(3, EPOLL_CTL_ADD, 7, {EPOLLOUT, {u32=7, u64=7}}) = 0
22:58:40.880728 gettimeofday({1554325120, 880760}, NULL) = 0
22:58:40.880827 epoll_wait(3, [{EPOLLOUT|EPOLLERR|EPOLLHUP, {u32=7, u64=7}}], 200, 1000) = 1
22:58:40.912014 gettimeofday({1554325120, 912089}, NULL) = 0
22:58:40.912216 getsockopt(7, SOL_SOCKET, SO_ERROR, [111], [4]) = 0
22:58:40.912601 setsockopt(7, SOL_SOCKET, SO_LINGER, {onoff=1, linger=0}, 8) = 0
22:58:40.913052 close(7)                = 0
```

Before the commit, we waited for the connect() call to complete (here's a reachable destination port, so that we can see when the sendto() appears):

```
23:01:28.744448 socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 7
23:01:28.744627 fcntl(7, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
23:01:28.744856 setsockopt(7, SOL_TCP, TCP_NODELAY, [1], 4) = 0
23:01:28.745058 setsockopt(7, SOL_TCP, TCP_QUICKACK, [0], 4) = 0
23:01:28.745233 connect(7, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("51.15.8.218")}, 16) = -1 EINPROGRESS (Operation now in progress)
23:01:28.745398 gettimeofday({1554325288, 745434}, NULL) = 0
23:01:28.745545 epoll_wait(3, [], 200, 1000) = 0
23:01:29.746921 gettimeofday({1554325289, 747024}, NULL) = 0
23:01:29.747403 getsockopt(7, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
23:01:29.747778 connect(7, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("51.15.8.218")}, 16) = 0
23:01:29.748067 sendto(7, "PING\r\n", 6, MSG_DONTWAIT|MSG_NOSIGNAL, NULL, 0) = 6
23:01:29.748285 recvfrom(7, 0xd65504, 16384, 0, NULL, NULL) = -1 EAGAIN (Resource temporarily unavailable)
23:01:29.748472 getsockopt(7, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
23:01:29.748653 setsockopt(7, SOL_SOCKET, SO_LINGER, {onoff=1, linger=0}, 8) = 0
23:01:29.748851 close(7)   
```

And here is a unreachable port before the commit, as we can see haproxy gives up because the connect() call does not return successfully and sento() is never called. In this case we also show the correct error (`Connection refused at step 1 of tcp-check (connect port 81)`):

```
22:57:23.905640 socket(PF_INET, SOCK_STREAM, IPPROTO_TCP) = 7
22:57:23.905761 fcntl(7, F_SETFL, O_RDONLY|O_NONBLOCK) = 0
22:57:23.905849 setsockopt(7, SOL_TCP, TCP_NODELAY, [1], 4) = 0
22:57:23.905954 setsockopt(7, SOL_TCP, TCP_QUICKACK, [0], 4) = 0
22:57:23.906051 connect(7, {sa_family=AF_INET, sin_port=htons(81), sin_addr=inet_addr("51.15.8.218")}, 16) = -1 EINPROGRESS (Operation now in progress)
22:57:23.906155 gettimeofday({1554325043, 906174}, NULL) = 0
22:57:23.906255 epoll_wait(3, [], 200, 1000) = 0
22:57:24.907630 gettimeofday({1554325044, 907731}, NULL) = 0
22:57:24.908089 getsockopt(7, SOL_SOCKET, SO_ERROR, [111], [4]) = 0
22:57:24.908497 setsockopt(7, SOL_SOCKET, SO_LINGER, {onoff=1, linger=0}, 8) = 0
22:57:24.908870 close(7)                = 0
```

This commit is in 1.7.11, 1.8.5 and 1.9-dev1+.

**Comment by wtarreau** [comment]
This one is tricky. While I can understand that from a user's perspective this is wrong, from a low-level perspective it's the correct way of doing it. We could possibly hack something and pretend there is nothing to send after the connect rule, so that we break out of the loop, but that doesn't appeal me much.

For similar reasons if there are multiple send rules in a row, they will be aggregated so that we don't send incomplete messages spanning over multiple packets, and in this case it will be the last send of the series which will report a write error.

At the very least the current behaviour must be properly documented so that the user always has a way to know what error happened.

Just thinking loud, maybe we should have a "barrier" or "break" operation to force to serialize certain operations ? Those willing to spot issues in finest details could place them after connect() or individual send(), and those who favor compatibility, performance, or limited alarms on the server side would not use them. That's just an idea of course.

**Comment by wtarreau** [comment]
I'm marking it "works as designed" since technically this is true, which doesn't mean I'm against changing the design if we find a better option!

**Comment by capflam** [comment]
This issue seems to be fixed now. I tested all versions since the 1.7 and the correct message is reported in logs. For the 2.0, it was fixed in the 2.0.2 by the commit 7199e96c.
